mkdocs/structure/pages.py

```diff
@@ -172,10 +174,20 @@ def _set_canonical_url(self, base: Optional[str]) -> None:
             self.canonical_url = None
             self.abs_url = None
 
-    def _set_edit_url(self, repo_url: Optional[str], edit_uri: Optional[str]) -> None:
-        if edit_uri:
+    def _set_edit_url(
+        self,
+        repo_url: Optional[str],
+        edit_uri: Optional[str] = None,
+        edit_uri_template: Optional[str] = None,
+    ) -> None:
+        if edit_uri or edit_uri_template:
             src_uri = self.file.src_uri
-            edit_uri += src_uri
+            if edit_uri_template:
+                noext = posixpath.splitext(src_uri)[0]
+                edit_uri = edit_uri_template.format(path=src_uri, path_noext=noext)
+            else:
+                assert edit_uri is not None and edit_uri.endswith('/')
```

Comment:
I usually don't like assertions like this in the code but since it's used some other places in this file, I guess I can live with it

Reply:
Assertions like what 😅
Also other assertions in this file are probably recently added by me 😅
Anyway my approach towards Python assert is "it can actually literally never happen"; this is not error checking. In this case it's to satisfy the type checker, it looks at assertions